### PR TITLE
[ELF] OVERLAY: support optional start address and LMA

### LINKED ISSUE #77265: [lld][ELF] OVERLAY: support optional start address and LMA
When I try to run binutils ld test with lld to figure out difference between gnu ld and lld, I found lld reports error:
```
error:  overlay-size.t:27: malformed number: :
>>>   OVERLAY : 
>>>           ^
```
where the linker script is
```
MEMORY
{
  TEXTMEM (ARX) : ORIGIN = 0x10000, LENGTH = 32K
  DATAMEM (AW)  : ORIGIN = 0x20000, LENGTH = 32K
  LOADMEM (AW)  : ORIGIN = 0x30000, LENGTH = 32K
}

/* Map should be:

           SIZE    VMA    LMA
   .bss1     10  20000  20000
   .bss2     30  20000  20010
   .bss3     20  20000  20040
   .mbss    230  20030  20060

   .mtext    20  10000  30000
   .text1    80  10020  30020
   .text2    40  10020  300a0
   .text3    20  10020  300e0

   .data1    30  20260  30100
   .data2    40  20260  30130
   .data3    50  20260  30170  */

SECTIONS
{
  OVERLAY : 
    {
      .bss1 { *(.bss1) }
      .bss2 { *(.bss2) }
      .bss3 { *(.bss3) }
    } > DATAMEM

  end_of_bss_overlays = . ;
    
  .mtext : { *(.mtext) } > TEXTMEM AT > LOADMEM

  .mbss : AT (__load_stop_bss3)
    {
      *(.mbss)
      . += 0x200;
    } > DATAMEM

  OVERLAY :
    {
      .text1 { *(.text1) }
      .text2 { *(.text2) }
      .text3 { *(.text3) }
    } > TEXTMEM AT > LOADMEM

  end_of_text_overlays = . ;

  OVERLAY :
    {
      .data1 { *(.data1) }
      .data2 { *(.data2) }
      .data3 { *(.data3) }
    } > DATAMEM AT > LOADMEM

  end_of_data_overlays = . ;
  
  . = 0x8000;
  /DISCARD/ : { *(.reginfo) *(.MIPS.abiflags) }
}
```
and the source file is
```
	.section .bss1, "aw", "nobits"
	.space 0x10
	.section .bss2, "aw", "nobits"
	.space 0x30
	.section .bss3, "aw", "nobits"
	.space 0x20

	.section .text1, "ax", "progbits"
	.space 0x80
	.section .text2, "ax", "progbits"
	.space 0x40
	.section .text3, "ax", "progbits"
	.space 0x20

	.section .data1, "aw", "progbits"
	.space 0x30
	.section .data2, "aw", "progbits"
	.space 0x40
	.section .data3, "aw", "progbits"
	.space 0x50

	.section .mtext, "ax", "progbits"
	.space 0x20
	.section .mbss, "aw", "nobits"
	.space 0x30
```
It looks lld treats symbol : as a tok, and this issue is similar with this : [https://discourse.llvm.org/t/memory-command-region-name-syntax/75145](https://discourse.llvm.org/t/memory-command-region-name-syntax/75145)
which lld treats string as Expr and split in function tokenizeExpr.

### PR DESCRIPTION
**Pull request by MaskRay**
https://reviews.llvm.org/D44780 implemented rudimentary support for
OVERLAY. The start address and `AT(ldaddr)` in `OVERLAY [start] :
[NOCROSSREFS] [AT ( ldaddr )]` are not optional.

In addition, there are two issues:

* When the start address is `.`, subsequent sections don't share the
  address of the first overlay section.
* When the first overlay section is empty and discardable, `p_paddr` is
  incorrectly zero. This is because a discarded section has a zero
  address, causing `prev->getLMA() + prev->size` where `prev` refers to
  the first section to evaluate to zero.

This patch supports optional start address and LMA and fix the issues.
Close #77265

### CODE REVIEW

File: lld/ELF/ScriptParser.cpp [hunk 1]
```diff
@@ -547,10 +549,15 @@ SmallVector<SectionCommand *, 0> ScriptParser::readOverlay() {
...
-    if (prev)
+    if (prev) {
       osd->osec.lmaExpr = [=] { return prev->getLMA() + prev->size; };
-    else
+    } else {
       osd->osec.lmaExpr = lmaExpr;
+      // Ensure that the first section, even if empty, is not discarded.
+      // Subsequent sections share its address.
+      osd->osec.usedInExpression = true;
+      addrExpr = [=]() -> ExprValue { return {&osd->osec, false, 0, ""}; };
```
Comment (smithp35) [inline lld/ELF/ScriptParser.cpp:559]: I guess addrExpr is for the case where the original addrExpr is DOT. May be worth updating the comment to.
```
// Use first section address for subsequent sections as initial addrExpr can be DOT. 
```

**Review by smithp35 — APPROVED** [review]
A couple of comment suggestions, otherwise LGTM

File: lld/ELF/ScriptParser.cpp (hunk 1 above)
Comment (MaskRay) [reply lld/ELF/ScriptParser.cpp:561]: Thanks for the suggestion. Adopted!

Comment (MaskRay) [reply lld/ELF/ScriptParser.cpp:561]: Thanks for the suggestion. I reworded it a bit.
```
+  // When AT is omitted, LMA should equal VMA. script->getDot() when evaluating
+  // lmaExpr will ensure this, even if the start address is specified.
```

Setting `lmaExpr` to `script->getDot()` ensures that LMA propagation in `LinkerScript::assignOffsets` is disabled and `state->lmaOffset = sec->lmaExpr().getValue() - dot;` evaluates to 0.

**Comment by MaskRay** [comment]
Manually merged 7c89b20e02ff079ec84fc54880dbc6c063d8c915. Hmm, it seems that my `spr diff` workflow breaks again.
https://discourse.llvm.org/t/using-getcord-spr/76097/3

**Pull request closed without merging**